### DISCUSSION TITLE
404 and 403 errors

### DISCUSSION THREAD

**Discussion by hahahahywel** [Chimildic/goofy#175] [Помощь с ошибкой] [answered]
Hi there,

I keep getting these errors for my playlist. Funny thing is it was working fine a few days ago but has recently kept on failing. I went as far as to reinstall Goofy but the problem persists. This is part way down the list of errors showing both 404 and 403 error codes.

![Screenshot 2022-03-20 105329](https://user-images.githubusercontent.com/72909551/159158922-7394c2d7-a097-4f9c-a1c5-e51df7f8cabb.png)

When debugging the code, despite having all these errors the correct array is there but it is only saving 600 or so tracks.

Here is the code:

```
function TenKHist() {

 let play_hist = Source.getTracks([
    {name: 'Play History', id:'06VIxKZ3cCEyd5OmFSt6ji'},
    {name: 'Play History archive 1', id:'6Mtlf3DLANK1sxcEtlO1XQ'},
  ]);

 Selector.keepFirst(play_hist,10000);


  // add to playlist
  Playlist.saveWithReplace({
    id: '38hG1BYIwYRunoOBEHSXlh',
    name: 'Recent 10k played',
    tracks: play_hist,
  });


}
```
Any help would be appreciated
Thanks

**Comment by dub4u** [comment]
Same for me. First time I saw it was March ~~20~~ 18. Then what happens is that `Source.getTracks()` doesn't contain all tracks of the playlist, those where the API 404'd are missing.

**Comment by ladywhiskers** [comment]
I've had similar errors since 18 March 2022, not every time though, and it doesn't appear with any consistency.

**Comment by hahahahywel** [comment]
Whatever changes have been made seem to have fixed the problem! Maybe it was an issue with the spotify API but whatever the issue was the playlist now seems to work properly.

Thanks for a great program :)

**Comment by efremych** [comment]
Тоже зачистила ошибка 403.
Уже пару раз точно за неделю (из того что замечал). Итого в плейлисте 200 треков вместо 365. Начинаю переживать за плейлисты, треки в которых накапливаются вручную и не только - эдак они и обнулиться могут :(

![1E392F03-0DB0-4BA6-8320-83C0957748AF](https://user-images.githubusercontent.com/76212904/171230184-16f7fa87-c543-45f8-8330-4f99382b1d05.jpeg)

**Reply by Chimildic** [reply]
Во первых, лог говорит что собралось 365 треков. Почему 200?
Во вторых, прямой запрос плейлиста не дает ошибок. Нужны какие-то условия для воспроизведения. 
Возможно это другое воплощение [бага со значением total](https://github.com/Chimildic/goofy/discussions/183#discussioncomment-2625371) с их стороны. Оно обновляется с лагом во времени. Поэтому goofy мог получить неактуальное значение и просто отправить лишний запрос.

**Reply by efremych** [reply]
Фактически 200 треков попало в плейлист (saveWithReplace). Что я не сразу заметил, так как в автоматическом моем описании плейлиста указано 365 :)
Добавляются же треки партиями? Получается не все запросы на добавление треков смогли отработать успешно…

**Reply by efremych** [reply]
![CBCB15F0-87C0-400F-AE57-36BCB603174F](https://user-images.githubusercontent.com/76212904/172302852-880bd7f4-468e-4e3a-9a75-7153215fdb5a.png)

опять 200 треков

**Comment by efremych** [comment]
> эдак они и обнулиться могут

так и случилось в итоге :)
Есть плейлист, в который автоматически дублируются Любимые треки. 
Сравнивается размер массива любимых треков и размер массива плейлиста-дублёра. Если размеры не равны, то плейлист обновляется. Каждый день такая проверка. 
Видимо запрос любимых треков (не через кэш) дал пустой массив, условие выполнилось и так как любимых треков, якобы, 0, то плейлист отчистился :)

никаких ошибок в логах нет (кроме ошибок других функций, вследствие пустого массива любимых треков)… 
скорее, для информации пишу, что такое может быть. Впервые за 1,5 года.

**Reply by Chimildic** [reply]
Любимые треки запрашиваются через дубль-плейлист? Почему не напрямую `getSavedTrack`?

Ко мне тоже приходила ошибка _index out of bounds_ при использовании `saveWithReplace`. Но как пришла, так и ушла. Ничего не изменял, поймать и воспроизвести не удалось. 

Учитывая сколько времени такой ошибки не существовало да и отсутствие изменений в этой части кода, только на сам Spotify пожаловаться хочется.

**Reply by Chimildic** [reply]
Если подставить недоступное _id трека_, например, `1WjYpzuSuoeZQ0zjvDPUBl`. Ошибка _index out of bounds_ будет. **Но** до нее приходит другая ошибка - _payload contains a non-existing ID_, которая поясняет что пошло не так. В случаях же, описанных в этой теме, поясняющих ошибок нет. Сразу ошибка про индекс. 

То есть можно предположить, Spotify по _неназванной_ нам причине отказался добавлять минимум 1 трек, что ломает индекс для следующих запросов. И в то же время, повторный запуск с тем же набором треков не вызывает ошибок.

**Reply by efremych** [reply]
> Любимые треки запрашиваются через дубль-плейлист? Почему не напрямую getSavedTrack?

Нет, любимые треки запрашиваются стандартно. Просто держу также их и в виде автоматически синхронизируемого плейлиста, в котором также есть и локальные треки. Так и заметил неладное - слушал себе этот плейлист и вдруг начали играть только локальные треки из него, оказалось, что так вот отработал скрипт :)

> Учитывая сколько времени такой ошибки не существовало да и отсутствие изменений в этой части кода, только на сам Spotify пожаловаться хочется.

В данном случае в логах GS, как уже говорил, пусто, но склоняюсь, что это всё из одной серии... Хоть и в одном случае (с любимыми треками) как будто ошибка чтения, а в другом - записи.
На этой неделе не замечал, чтобы не все треки из массива добавлялись в плейлист, но на прошлой - чуть ли не через день было (200 или 300 треков из 365 добавлялись в плейлист, как при автоматической отработке скрипта по тригеру, так и при ручном запуске).

Я уж думал, что смена Региона как-то повлияла...

**Reply by efremych** [reply]
Только добрался до компа, только скопировал всё обратно в плейлист в нормальном порядке следования треков, только добавил условие в скрипт, чтобы если массив пустой, то не переслхранять треки.  Как опять, было 581 треков, стало 531 (минус ровно 50) :) Массив и не пустой, но и не полный)

**Reply by Chimildic** [reply]
Снова _index out of bounds_ без доп описания в предыдущих сообщениях? 

Давайте попробуем посмотреть на _id треков_, когда происходит ошибка. У функции `modifyTracks` поменять одно из условий на следующее
```js
if (requestType === 'post') {
    // post-запрос добавляет треки в плейлист
    let response = SpotifyRequest.post(url, payload);
    console.log(`payload length = ${payload.uris.length}`)
    if (response == undefined) {
      console.log(`begin = ${begin}, end = ${end}, payload: \n${payload.uris}`);
    }
}
```

**Reply by efremych** [reply]
В логах GS никаких сообщений об ошибках.
Спасибо, попробую, добавлю, послежу. Правда через недельку только…

**Comment by Chimildic** [comment]
Один из пользователь поделился скриптом, который относительно часто приводит к ошибке _index out of bounds_. Она возникает потому что Spotify буквально сломан.

Для теста добавил проверку количества треков после _каждого_ запроса. В результате всего треков: _100, 200, 900, 400_ или ... _300, 400, 500, 400, 700._ Была операция замены. Будто у них синхронизации нет. Из-за того, что Spotify не может определиться какое количество треков сейчас в плейлисте, наши запросы с указанием позиции отклоняли ошибкой 403.

В версии goofy 1.8.2 изменил алгоритм добавления треков. Теперь позиция в запросе отправляется только когда явно указано добавление треков в начало плейлиста. По умолчанию треки добавляются в конец. Тогда вычисление позиции вставки это проблема для Spotify.

- Сейчас будут появляться ошибки 500 и 502 с попыткой повторной отправки. Они говорят о _внутренних проблемах на стороне Spotify_. Полагаю как раз из-за рассинхрона - не понимает куда треки вставить. 
- При использовании `saveWithAppend` или `saveWithUpdate` с указанием `position = 'begin'` по прежнему будут возникать ошибки _403 index out of bounds_. Также будут предприняты попытки повторной отправки.
- В крайнем случае добавление прерывается. На Google Диске создается резервная копия, чтобы избежать потери треков.

Пока будет удачное завершение, ошибки можно игнорировать.

**Reply by Bondian** [reply]
Я правильно понимаю, что после обновления библиотеки до 1.8.2 надо всё перелопатить и заменить во всех плейлистах "toEnd: true" на "position: end"?
Чёрт, туплю..., По умолчанию "position: end", значит просто удалить "toEnd: true", чтобы ошибок не было?

**Reply by Chimildic** [reply]
Присутствие _toEnd_ не вызовет ошибок.
Где важно добавлять в начало нужно дописать _position_.

**Reply by Bondian** [reply]
Ясно.

**Reply by Intuki** [reply]
> Один из пользователь поделился скриптом, который относительно часто приводит к ошибке _index out of bounds_. Она возникает потому что Spotify буквально сломан.

В версии Goofy 1.8.1 скрипт, формирующий большой плейлист на основе истории прослушанного:
```js
function genRecentTracks(){
    let recentTracks = RecentTracks.get(3500);
    
    Playlist.saveWithReplace({
        name: '3500 Recent Tracks',
        tracks: recentTracks,
        randomCover: 'update',
    });
}
```
 часто (но не всегда) заканчивался с ошибками 403.

Версия 1.8.2, несмотря на ошибки 502, благодаря повторным отправкам регулярно обеспечивает правильный результат.

---

Еще любопытно, что до этого в версии 1.8.1 сталкивался с тем, что использование одного большого плейлиста через:
```js
    let tracks = Source.getTracks([
        { id: 'id-огромного-плейлиста' },
    ]);
```
приводило к гораздо большей вероятности появления проблем с  ошибками 403 "Index out of bounds", чем использование через:
```js
    let tracks = Source.getPlaylistTracks('', 'id-огромного-плейлиста');
```

**Reply by Chimildic** [reply]
Какое-то время назад ошибки 403 не существовало даже с большими плейлистами. Но они что-то сломали у себя.

Разницы между `getPlaylistTracks` и `getTracks` нет. Первая внутри себя вызывает вторую.

**Reply by Intuki** [reply]
> Какое-то время назад ошибки 403 не существовало даже с большими плейлистами. Но они что-то сломали у себя.
> 
> Разницы между `getPlaylistTracks` и `getTracks` нет. Первая внутри себя вызывает вторую.

Да, согласен, от этого не зависит. Похоже на стороне Spotify иногда один и тот же скрипт отрабатывает нормально, а иногда можно запустить несколько раз, и все запуски будут с ошибками.

**Comment by Chimildic** [comment, marked as answer]
Поддержка Spotify ответила про ошибки 500+ при создании плейлистов. 

- For playlists that are frequently updated, the playlist state becomes momentarily inconsistent. However, it should be fixed within seconds.
- Meanwhile, the 500 Internal Server Error appeared because of a recent migration. This has been completed and performance is way better now.

Как я и [предполагал раньше](https://github.com/Chimildic/goofy/discussions/175#discussioncomment-3033687), происходит рассинхрон. В общем случае, повторная отправка запроса должна помочь (она как раз происходит после паузы).

**Reply by tupizm** [reply]
То есть теперь с этим жить будем?

**Reply by Chimildic** [reply]
@tupizm Ваши скрипты не справляются сейчас?

**Reply by tupizm** [reply]
Сейчас проверил утреннее выполнение. Вроде без ошибок. Будем следить. А так, я понимаю, только разбивать на мелкие, чтобы в случае повторных запросов времени хватило отработать.

**Reply by Chimildic** [reply]
@tupizm Насколько помню, там приличный объем. Так что разбить на части не самый плохой вариант. Также можно поиграть с паузами, уменьшив простой. В крайнем случае делать отложенный триггер, который создается программно на один запуск. Но надеюсь до этого не дойдет.
